apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs

```diff
@@ -0,0 +1,474 @@
+                // So use the manually configured projectId and override the resource.
+                projectId = GaxPreconditions.CheckNotNull(ProjectId?.Render(LogEventInfo.CreateNullEvent()), nameof(ProjectId));
+                if (ResourceType == null)
+                {
+                    resource = new MonitoredResource { Type = "global", Labels = { { "project_id", projectId } } };
+                }
+                else
+                {
+                    resource = new MonitoredResource { Type = ResourceType,
+                        Labels = { ResourceLabels.ToDictionary(x => x.Name, x => x.Layout.Render(LogEventInfo.CreateNullEvent())) } };
+                }
+            }
+
+            _resource = resource;
+            var logName = new LogName(projectId, logId);
+            _logName = logName.ToString();
+            _logNameToWrite = LogNameOneof.From(logName);
+        }
+
+        /// <summary>
+        /// Writes async log event to the log target.
+        /// </summary>
+        /// /// <param name="logEvent">Logging event to be written out.</param>
+        protected override void Write(AsyncLogEventInfo logEvent)
+        {
+            LogEntry logEntry = null;
+
+            try
+            {
+                logEntry = BuildLogEntry(logEvent.LogEvent);
+            }
+            catch (Exception ex)
+            {
+                InternalLogger.Error(ex, "GoogleStackdriver(Name={0}): Failed to create LogEntry, marked as failed", Name);
+                logEvent.Continuation(ex);
+            }
+
+            if (logEntry != null)
+            {
+                WriteLogEntries(new[] { logEntry }, logEvent.Continuation);
+            }
+        }
+
+        /// <summary>
+        /// Writes a list of logging events to the log target.
+        /// </summary>
+        protected override void Write(IList<AsyncLogEventInfo> logEvents)
+        {
+            List<LogEntry> logEntries = new List<LogEntry>(logEvents.Count);
+            List<AsyncContinuation> continuationList = new List<AsyncContinuation>(logEvents.Count);
+            for (int i = 0; i < logEvents.Count; ++i)
+            {
+                var logEvent = logEvents[i];
+
+                try
+                {
+                    var logEntry = BuildLogEntry(logEvent.LogEvent);
+                    logEntries.Add(logEntry);
+                    continuationList.Add(logEvent.Continuation);
+                }
+                catch (Exception ex)
+                {
+                    InternalLogger.Error(ex, "GoogleStackdriver(Name={0}): Failed to create LogEntry, marked as failed", Name);
+                    logEvent.Continuation(ex);
+                }
+            }
+
+            if (logEntries.Count > 0)
+            {
+                WriteLogEntries(logEntries, continuationList);
+            }
+        }
+
+        private void WriteLogEntries(IList<LogEntry> logEntries, object continuationList)

```

Comment:
I don't quite understand the logic behind this throttling.
It appears that it'll wait one second, but if there are still too many tasks pending, it'll continue anyway; is that correct?
Could it wait until `_pendingTaskCount` is less than `TaskPendingLimit`? (with some kind of timeout, which will cause failure?)
To do this properly, I suspect a list/queue of Tasks will need to be kept, rather than just a single_prevTask.
What do you think?

Please can you also add a new boolean, or refactor the code some other way, so that `belowTaskLimit` doesn't change meaning.

And please comment the final code, so readers, ie me ;) , can understand it.

This task-limit code will also need testing.

I'm happy to do any/all of this if you'd like. Please let me know.

Reply:
Yes if the `_prevTask` has gone bad, then it just starts a new task-chain ignoring the one that doesn't complete by itself.

I have now made the timeout a little more aggressive, so if throttle is required, then it doesn't wait for all pending tasks to complete, but just waits until below the pending task limit.

Not sure I understand the case with `belowTaskLimit` changing meaning. Please explain.

Reply:
@chrisdunelm Should I make the timeout of 1.5 sec configurable, so users can configure longer delays for slower internet ?

Reply:
Sorry, my mistake about `belowTaskLimit`. I had misread the code slightly.
Yes, I agree it would be a good idea to make the timeout configurable; but I think it's time to get this PR merged. I'll create an issue with the remaining work to do before release, including this timeout configuration.